Add error handlers to code generation

chengluyu opened the pull request:
This PR adds error handlers to code generation. So if code generation crashes, it can still displayed inferred types.

chengluyu committed:
Wrap code generation with `try`-`catch` and show errors if has

LPTK commented on js/src/main/scala/Main.scala:
```diff
@@ -138,13 +104,89 @@ object Main {
     )
   }
 
+  // Extract a separated function to do this.
```
```suggestion
```

LPTK replied:
This comment line is pretty useless 😄

chengluyu replied:
Removed. 😄

LPTK commented on js/src/main/scala/Main.scala:
```diff
@@ -138,13 +104,89 @@ object Main {
   }
 
+  // Extract a separated function to do this.
+  // Returns `Left[Str]` if successful, `Right[Str]` if not.
```
Note that the convention is `Right` for successful (it also sounds better, being able to say "the _right_ result").

chengluyu replied:
Changed.

LPTK commented on js/src/main/scala/Main.scala:
```diff
@@ -289,14 +331,24 @@ object Main {
           declared.get(nme).foreach { sign =>
             // ctx += nme -> sign  // override with less precise declared type?
-            subsume(ty_sch, sign)(ctx, raise, TypeProvenance(d.toLoc, "def definition"))
+            subsume(ty_sch, sign)(
```
Same remark as in the other PR.

chengluyu replied:
Changed.

LPTK commented on js/src/main/scala/Main.scala:
```diff
@@ -289,14 +331,24 @@ object Main {
           val errProv = TypeProvenance(rhs.toLoc, "def signature")
-          val ty_sch = PolymorphicType(0, typeType(rhs)(ctx.nextLevel, raise,
-            tps.map(tp => tp.name -> freshVar(noProv/*FIXME*/)(1)).toMap))
+          val ty_sch = PolymorphicType(
```
Same remark as in the other PR.

chengluyu replied:
Reverted.

LPTK reviewed (approved):
Ok, LGTM. Please go ahead and merge after making the minor changes. Thanks!

chengluyu committed:
Adopt suggestions in the PR

1. Revert changes made by formatter.
2. Remove useless comments.
3. Use `R(code)` instead of `L(code)`
